File: BaekjoonAlgorithm/BA_11049/BJAlgo_11049.py
from sys import stdin

n = int(stdin.readline())
matrix = []
for _ in range(n):
    matrix.append(list(map(int, stdin.readline().split())))

dp = [[0] * n for _ in range(n)]
for i in range(n - 1):
    dp[i][i + 1] = matrix[i][0] * matrix[i][1] * matrix[i + 1][1]


def find_add_num(start, end, mid):
    return matrix[start][0] * matrix[mid][0] * matrix[end][1]


for j in range(2, n):
    i = 0
    while i + j < n:
        for k in range(i, i + j):
            if dp[i][i + j] == 0:
                dp[i][i + j] = dp[i][k] + dp[k + 1][i + j] + find_add_num(i, i + j, k + 1)
            else:
                dp[i][i + j] = min(dp[i][i + j], dp[i][k] + dp[k + 1][i + j] + find_add_num(i, i + j, k + 1))
        i += 1

print(dp[0][n - 1])

File: BaekjoonAlgorithm/BA_11049/BJAlgo_11049.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;

public class BJAlgo_11049 {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        int[][] matrix = new int[n][2];
        for (int i = 0; i < n; i++) {
            matrix[i] = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        }

        int[][] dp = new int[n][n];
        for (int i = 0; i < n - 1; i++) {
            dp[i][i + 1] = matrix[i][0] * matrix[i][1] * matrix[i + 1][1];
        }

        for (int j = 2; j < n; j++) {
            for (int i = 0; i + j < n; i++) {
                for (int k = i; k < i + j; k++) {
                    if (dp[i][i + j] == 0) {
                        dp[i][i + j] = dp[i][k] + dp[k + 1][i + j] + (matrix[i][0] * matrix[k + 1][0] * matrix[i + j][1]);
                    } else {
                        dp[i][i + j] = Math.min(dp[i][i + j], dp[i][k] + dp[k + 1][i + j] + (matrix[i][0] * matrix[k + 1][0] * matrix[i + j][1]));
                    }
                }
            }
        }

        System.out.println(dp[0][n - 1]);
    }
}
